File: archives/cpp_gui/menu/workspace/workspace_merge_into.cpp
#include "workspace_merge_into.h"


MenuWorkspaceMergeInto::MenuWorkspaceMergeInto(QWidget* parent) : 
    IodeSettingsDialog(parent)
{
    setupUi(this);

    QList<QString> qIodeType;
    for(const std::string& iode_type: v_iode_types) qIodeType << QString::fromStdString(iode_type);

    wComboBoxObjectType = new WrapperComboBox(label_object_type->text(), *comboBox_object_type, REQUIRED_FIELD, qIodeType);
    wFileChooserInputFile = new WrapperFileChooser(label_input_file->text(), *fileChooser_input_file, REQUIRED_FIELD, FILE_COMMENTS, EXISTING_FILE);

    mapFields["ObjectType"] = wComboBoxObjectType;
    mapFields["InputFile"] = wFileChooserInputFile;

    connect(&wComboBoxObjectType->getQField(), &QComboBox::currentIndexChanged, this, &MenuWorkspaceMergeInto::updateFileChooser);

    className = "MENU_WORKSPACE_MERGE_INTO";
    loadSettings();
}

MenuWorkspaceMergeInto::~MenuWorkspaceMergeInto()
{
    delete wComboBoxObjectType;
    delete wFileChooserInputFile;
}

void MenuWorkspaceMergeInto::merge_into_workspace()
{
    try
    {
        int iodeType = wComboBoxObjectType->extractAndVerify();
        std::string input_file = wFileChooserInputFile->extractAndVerify().toStdString();

        switch (iodeType)
        {
        case COMMENTS:
            Comments.merge_from(input_file);
            break;
        case EQUATIONS:
            Equations.merge_from(input_file);
            break;
        case IDENTITIES:
            Identities.merge_from(input_file);
            break;
        case LISTS:
            Lists.merge_from(input_file);
            break;
        case SCALARS:
            Scalars.merge_from(input_file);
            break;
        case TABLES:
            Tables.merge_from(input_file);
            break;
        case VARIABLES:
            Variables.merge_from(input_file);
            break;
        default:
            break;
        }
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }

    this->accept();
}

void MenuWorkspaceMergeInto::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/iode_gui/menu/workspace/workspace_merge_into.py
from qtpy.QtCore import Slot
from qtpy.QtWidgets import QWidget, QMessageBox

from iode_gui.settings import MixinSettingsDialog
from iode_gui.abstract_main_window import AbstractMainWindow
from .ui_workspace_merge_into import Ui_MenuWorkspaceMergeInto


from iode import (IodeType, IodeFileType, comments, equations, identities, 
                  lists, scalars, tables, variables)


class MenuWorkspaceMergeInto(MixinSettingsDialog):
    def __init__(self, parent: AbstractMainWindow):
        super().__init__(parent)
        self.ui = Ui_MenuWorkspaceMergeInto()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.v_files_types = list(IodeFileType)
        self.v_iode_types = list(IodeType)
        v_iode_type_names = [iode_type.name for iode_type in self.v_iode_types]
        self.ui.comboBox_object_type.addItems(v_iode_type_names)    
        self.ui.comboBox_object_type.currentIndexChanged.connect(self.update_file_chooser)

        self.load_settings()

    @Slot()
    def merge_into_workspace(self):
        try:
            iode_type: IodeType = self.v_iode_types[self.ui.comboBox_object_type.currentIndex()]
            input_file: str = self.ui.fileChooser_input_file.filepath

            if not len(input_file):
                raise ValueError("'input file' is empty")

            if iode_type == IodeType.COMMENTS:
                comments.merge_from(input_file)
            elif iode_type == IodeType.EQUATIONS:
                equations.merge_from(input_file)
            elif iode_type == IodeType.IDENTITIES:
                identities.merge_from(input_file)
            elif iode_type == IodeType.LISTS:
                lists.merge_from(input_file)
            elif iode_type == IodeType.SCALARS:
                scalars.merge_from(input_file)
            elif iode_type == IodeType.TABLES:
                tables.merge_from(input_file)
            elif iode_type == IodeType.VARIABLES:
                variables.merge_from(input_file)

            self.accept()
        except Exception as e:
            self.parent().display_output(f"ERROR -> {str(e)}")
            QMessageBox.warning(self, "WARNING", "Failed to merge into workspace:\n" + str(e))

    @Slot(int)
    def update_file_chooser(self, i_iode_type: int):
        self.ui.fileChooser_input_file.enum_file_type = self.v_files_types[i_iode_type]
